### ISSUE TITLE
problems with panels bought in 2018 december

### ISSUE THREAD

**Issue by fotosettore** [2dom/PxMatrix#53]
hi
it seems something wrong in "new" panels 64x32
i bought first series of panels from a seller and no one works fine : random pixel on - no images - no text.
pattern test and pixeltime both not working.
i made about 10 building changing cables, panels, d1.mini and nodemcu.
i changed all possibily software setup (zigzag, line, 1/16 etc etc) . no luck
i think that new panel series, after 2018 october, changed in something, so them do not work with the available (oct 2018) pxmatrix library.
i sent a my panels to a user that has a panel that works fine and confirm to me than they do no work.
i saw other users complain about the same problem.
today were delivered panels form other seller : same problem.
any help ?

**Comment by 2dom** [comment]
Have you ensured that there is a proper ground connection between ESP and panel? It may also be a timing issue..

**Comment by fotosettore** [comment]
yes, i had.
anyhow the ground between power supply of panel and all ground pins of all connector are electrically connected together. so I do not think that's the problem. however i tried to connect ground of power supply to ground of nodemcu. same problem.
the same sketch works fine on 64x64 panels. there are same wiring connection.
please note i'm not only maker that complain about this problem : all last messages of last makers.
the makers before october does not have this trouble.
really very strange.

p.s. i removed some flickering problems using a prototype pcboard made with my cnc. this means that wiring connection must be as short as possible.
![img-20181228-wa0085](https://user-images.githubusercontent.com/45329325/50528957-c4ffb800-0af1-11e9-9dba-da6cbdb6de40.jpg)

**Comment by 2dom** [comment]
you could try reducing the SPI speed in the library...

**Comment by fotosettore** [comment]
SPI.setFrequency(20000000); 
this one ?

**Comment by 2dom** [comment]
yes ... that's the one. The 20 MHz might be too much...

**Comment by fotosettore** [comment]
I will try and let you know. now I do not have the 32 panel at hand.
many thanks for your support.

**Comment by fotosettore** [comment]
tried 
SPI.setFrequency(20000000)
SPI.setFrequency(2000000)
SPI.setFrequency(10000000)
SPI.setFrequency(1000000)
SPI.setFrequency(100000)
SPI.setFrequency(40000000)
SPI.setFrequency(4000000)

unlucky tests.
in SPI.setFrequency(1000000) brightness is low

**Comment by houk001** [comment]
I have the same problem with panels 64x32 ordered 11 dec from sryled display
https://www.aliexpress.com/item/P3-RGB-pixel-panel-HD-display-64x32-dot-matrix-p3-smd-rgb-led-module/32728985432.html
2 panels 3 nodemcu, 2 power supplies all the same, even pattern test not work as expected.

**Comment by fotosettore** [comment]
this is a half confirm that something changed in new panels :-(

**Comment by 2dom** [comment]
Ok ... could you upload a picture how it looks like?

**Comment by fotosettore** [comment]
this is my "effect"

![20181229_225128](https://user-images.githubusercontent.com/45329325/50542383-c2b66000-0bbc-11e9-9740-cd1fe97f8f06.jpg)

**Comment by 2dom** [comment]
what is it supposed to show?

**Comment by fotosettore** [comment]
pixeltime.ino

**Comment by 2dom** [comment]
which part? The weather icons and the color bars?

**Comment by houk001** [comment]
This is my random effect from pattern test Sometimes the upper line is white sometimes yellow but newer is going to the end of panel. Always some parts are black

**Comment by fotosettore** [comment]
i take the sketch pixeltime as is, 
i changed only
PxMATRIX display(64,32,P_LAT, P_OE,P_A,P_B,P_C,P_D,P_E);
and 
display.begin(16);

**Comment by 2dom** [comment]
Could you generate a video with the pattern test?

**Comment by fotosettore** [comment]
this is the pattern_test.ino with same needed changes

![20181229_230406](https://user-images.githubusercontent.com/45329325/50542470-63594f80-0bbe-11e9-8b8c-246aff304ad2.jpg)

**Comment by 2dom** [comment]
I need a video with one full cycle - red,yellow, white

**Comment by fotosettore** [comment]
i make video but NO CIRCLE ... all pixels are stopped

**Comment by fotosettore** [comment]
in ten minutes i make a video with 64x64 e 64x32.

**Comment by houk001** [comment]
For me pixels are moving line by line

 

Now it look like this:

 

https://cloud.hawke.pl/index.php/s/xpap2SQmJkgzaQ5

**Comment by houk001** [comment]
There is a movement in your video to. Line after some time go down by one pixel. On my both panels is go the same. 
I waited some time and red line goes to the half of size o panel and go up to the first line again.



Second blue line goes to the bottom of panel and goes up to the half of display again.

No other movement on panel.

**Comment by houk001** [comment]
Like this

**Comment by 2dom** [comment]
@fotosettore ok thanks ... that does not make much sense. 

Just a wild guess that either the latch or the output enable signal hast changed from activt low to active high or vice versa. Lets start with OE - could you try running the pattern test and modifying the latch function to this?

```
void PxMATRIX::latch(uint16_t show_time )
{
  
  digitalWrite(_LATCH_PIN,HIGH);
  digitalWrite(_LATCH_PIN,LOW);
  
  digitalWrite(_OE_PIN,HIGH); 
  delayMicroseconds(show_time);
  digitalWrite(_OE_PIN,LOW);
}
```

**Comment by fotosettore** [comment]
of course ... i try in few minutes

**Comment by houk001** [comment]
New video uploaded to cloud.

**Comment by 2dom** [comment]
@houk001 sorry ... cannot watch your videos ... loading forever :(

**Comment by fotosettore** [comment]
more lines of led on but does not works 
do you need a pic ?

**Comment by houk001** [comment]
Download it and then watch. VLC or similar is working

**Comment by 2dom** [comment]
ok ... lets try changing the latch pin polarity

```
void PxMATRIX::latch(uint16_t show_time )
{
  
  digitalWrite(_LATCH_PIN,LOW);
  digitalWrite(_LATCH_PIN,HIGH);
  
  digitalWrite(_OE_PIN,LOW); 
  delayMicroseconds(show_time);
  digitalWrite(_OE_PIN,HIGH);
}
```

**Comment by fotosettore** [comment]
now ... "half" works :-))
it needs video
please wait

**Comment by houk001** [comment]
Some progress :DDD video uploaded

**Comment by 2dom** [comment]
@houk001 Still cannot watch those videos ... download times 10min+

**Comment by fotosettore** [comment]
void PxMATRIX::latch(uint16_t show_time )
{  
  digitalWrite(_LATCH_PIN,LOW);
  digitalWrite(_LATCH_PIN,HIGH);
  
  digitalWrite(_OE_PIN,LOW); 
  delayMicroseconds(show_time);
  digitalWrite(_OE_PIN,HIGH);
}
made this effect:
http://www.homoandroidus.com/github/pxmatrix/VID-20181230-WA0001.mp4

**Comment by houk001** [comment]
https://youtu.be/fZrK874I-EU

**Comment by 2dom** [comment]
Try

```
void PxMATRIX::latch(uint16_t show_time )
{
digitalWrite(_LATCH_PIN,LOW);
digitalWrite(_LATCH_PIN,HIGH);
delayMicroseconds(10);
digitalWrite(_OE_PIN,LOW);
delayMicroseconds(show_time);
digitalWrite(_OE_PIN,HIGH);
delayMicroseconds(10);
}
```

**Comment by 2dom** [comment]
@houk001 Great ... that works! Very colorful :)

**Comment by fotosettore** [comment]
no changes with last void

**Comment by houk001** [comment]
Yes, no change with delays

**Comment by houk001** [comment]
A step back with first high. No more animated right to left movement.

**Comment by 2dom** [comment]
ok then ... another day. Cheers!

**Comment by fotosettore** [comment]
thanks for your support.
now problems is emerged i hope it will be solved .
have a nice night

**Comment by houk001** [comment]
Thanks from me too, i think that we are close to success :D

**Comment by 2dom** [comment]
Could you try

void PxMATRIX::latch(uint16_t show_time )
{
delayMicroseconds(100);
digitalWrite(_LATCH_PIN,LOW);
digitalWrite(_LATCH_PIN,HIGH);
delayMicroseconds(10);
digitalWrite(_OE_PIN,LOW);
delayMicroseconds(show_time);
digitalWrite(_OE_PIN,HIGH);
delayMicroseconds(10);
}

**Comment by houk001** [comment]
Of course, this is result 
https://youtu.be/pdtLYMSmJeY
first line goes only for half of the width but second goes better. on the end display goes black because of power supply 5v 3A that detect overload, I'm waiting for tomorrow to buy more powerful :)

**Comment by fotosettore** [comment]
no change for me too.
lines ok but other pixels low brightess
same result of http://www.homoandroidus.com/github/pxmatrix/VID-20181230-WA0001.mp4

**Comment by bribur** [comment]
With the latch and OE changes above mine is blank or has a few random lights for about 30 seconds then starts cycling and looks almost right, but after each complete cycle it moves down the page one line and there is ghosting.  I believe I have good grounding.  The E pin connection seems ambiguous, it is not grounded on the board yet is not driven by pxmatrix I think.  I have it floating currently.  Another clue noticed by others is that the chips along the edge changed from FM6124 to FM6126A and the supplier indicates the board is "upgraded".   But looking for FM6126A, a IR remote chip comes up so not sure what the change really is.  Mine has the FM6126A and I got it in December.

**Comment by 2dom** [comment]
@bribur That comment was very helpful ... I think FM6126A has a bit more logic build in ... will investigate ...

[FM6124.pdf](https://github.com/2dom/PxMatrix/files/2728582/FM6124.pdf)
[FM6126A.pdf](https://github.com/2dom/PxMatrix/files/2728583/FM6126A.pdf)

**Comment by houk001** [comment]
I also have FM6126A chip.

**Comment by 2dom** [comment]
Aparently the 6126A needs tree clock edges during latch ... can someone try this ?

```
void PxMATRIX::latch(uint16_t show_time )
{
  //digitalWrite(_OE_PIN,0); // <<< remove this
  digitalWrite(_LATCH_PIN,HIGH);
  digitalWrite(SPI_BUS_CLK,LOW);
  for (uint8_t latch_count=0; latch_count<3; latch_count++)
  {
    digitalWrite(SPI_BUS_CLK,HIGH);
    delayMicroseconds(1);
    digitalWrite(SPI_BUS_CLK,LOW);
    delayMicroseconds(1);
  }
  digitalWrite(_LATCH_PIN,LOW);
  digitalWrite(_OE_PIN,0); //<<<< insert this
  delayMicroseconds(show_time);
  digitalWrite(_OE_PIN,1);
}
```

**Comment by houk001** [comment]
![img_20190105_111542](https://user-images.githubusercontent.com/18401686/50723265-9c705300-10db-11e9-8cc4-fa79403c751d.jpg)
Only static display after this change, Random effect after reset.

**Comment by 2dom** [comment]
Reversed latch pin polarity ... can you try again ?

```
void PxMATRIX::latch(uint16_t show_time )
{
  //digitalWrite(_OE_PIN,0); // <<< remove this
  digitalWrite(_LATCH_PIN,LOW);
  digitalWrite(SPI_BUS_CLK,LOW);
  for (uint8_t latch_count=0; latch_count<3; latch_count++)
  {
    digitalWrite(SPI_BUS_CLK,HIGH);
    delayMicroseconds(1);
    digitalWrite(SPI_BUS_CLK,LOW);
    delayMicroseconds(1);
  }
  digitalWrite(_LATCH_PIN,HIGH);
  digitalWrite(_OE_PIN,0); //<<<< insert this
  delayMicroseconds(show_time);
  digitalWrite(_OE_PIN,1);
}
```

**Comment by houk001** [comment]
https://youtu.be/3-5ALrJa6JA Now it working like this :D

**Comment by 2dom** [comment]
can you run the pixeltime example and see how it looks?

**Comment by houk001** [comment]
Changed in pixeltime
PxMATRIX display(64,32,P_LAT, P_OE,P_A,P_B,P_C,P_D,P_E);
and in setup 
display.begin(16);

Result https://youtu.be/L9G4oel-X-g

**Comment by houk001** [comment]
https://pastebin.com/cq45nD9e

**Comment by houk001** [comment]
Tried all 3 LINE, ZIGZAG, ZAGGIZ and no success.

**Comment by 2dom** [comment]
Can you comment out setFastUpdate(true) through the sketch?

**Comment by houk001** [comment]
I only find one in setup.
After comment it look like this
![img_20190105_170739](https://user-images.githubusercontent.com/18401686/50726397-c42bdf00-110c-11e9-8bbd-ba48e9c642c0.jpg)

**Comment by bribur** [comment]
Yes, no doubt.  I can pass that along to them through Aliexpress and they can contact you directly if its ok to send them your email address shown in your profile.  Don't think it's practical to send you my panel, shipping would cost more than the panel.

**Comment by fotosettore** [comment]
i'm available to send my panels. please read private mail

**Comment by DaveDavenport** [comment]
Hit the same issue, and getting similar results.  IF there is anything I can do to debug, let me know.

**Comment by fotosettore** [comment]
my two panels were send Yesterday.
Dominic's patience and skill will do the rest ;-)

**Comment by DaveDavenport** [comment]
> my two panels were send Yesterday.
> Dominic's patience and skill will do the rest ;-)

Thanks! great.   (first time using these panels, cannot wait to start using them. )

**Comment by carlhako** [comment]
another confirmation here, 2 work colleges and I all ordered displays around early december from seller mentioned above. none of us could get them working with very similar results in pictures above, we also tried various other libraries for esp32, raspberry pi, purchased some rpi hats etc and nothing worked.

**Comment by 2dom** [comment]
Hi guys .... the last test looks rather promising
> https://youtu.be/3-5ALrJa6JA Now it working like this :D

I will give it a shot ...

**Comment by electric1490** [comment]
Hey @2dom , nice work getting that panel working! Really looking forward to learn how you managed to get that panel running. I have a feeling there are a lot of folks with DOA panels at the moment...

**Comment by 2dom** [comment]
Patience... haven't received the panels yet ... that test was performed by @houk001 using the latest edit. Does not translate into a working panel yet (see above)

**Comment by DaveDavenport** [comment]
I get the same result as above video with the 3 cycle  fix.

**Comment by cmc19** [comment]
It's worth mentioning that I'm having the same problem. I got my panel off amazon. Same FM6126A chips. 

https://www.amazon.com/panels-digital-module-display-P3-19296mm/dp/B079JSKF21/ref=as_li_ss_tl?ie=UTF8&qid=1524991179&sr=8-2&keywords=p3+led+matrix&dpID=51tf0f2jh3L&preST=_SY300_QL70_&dpSrc=srch&linkCode=sl1&tag=blough-20&linkId=8aeca1190b903eadd6136b76b3f8d9bf

**Comment by Kweck** [comment]
Same here. I just wasted 2 days following the instructions and searching for what I did wrong until I recognized this thread ;)
I owe this kind also with the FM6126A chip
https://www.amazon.de/gp/product/B07D5RFGMF/ref=ppx_yo_dt_b_asin_title_o03__o00_s00?ie=UTF8&psc=1

**Comment by DaveDavenport** [comment]
I don't understand exactly the story, but here https://github.com/hzeller/rpi-rgb-led-matrix/issues/746#issuecomment-453860510 is mentioned he got these panels working (well that he sais in 745 where he links to this comment).

**Comment by 2dom** [comment]
@DaveDavenport thanks for the heads up...wrote a few lines in reply to that comment...let's see what comes back

**Comment by DaveDavenport** [comment]
Giving the difference of the wiring (chaining back from output port to input) in PxMatrix versus how rpi-rgb-led-matrix is wired.  I am not sure how to rewrite that script exactly, would need to dive into it more.

Any hints?

**Comment by 2dom** [comment]
You could check out the latest master and use

`display.setDriverChip(FM6126A); `

after `display.begin(16).` Not doing anything yet but feel free to play with it ...

**Comment by shades66** [comment]
Hi,  I see you have already converted my script into PxMatrix,  I'm not going to get much chance to look till the weekend but as my original plan was to get the panels working with an ESP32 rather than an RPi  this seems like a good time to go back and try this library :)

One question questions,  you have the HIGH/LOW for clock & latch inverted  (ie, last 12/13 bits you have a digitalWrite(_LATCH_PIN,LOW); but in my script it was to set it HIGH),  is there a reason for that?

**Comment by 2dom** [comment]
@shades66 I did some measurement and digitalWrite appears to be inverted on the SPI clock pin after SPI is initialized. I also played with the polarity for the LATCH pin since the code did not do much at that point. You may want to comment out my latch code for FM6126A in PxMATRIX::latch() (@DaveDavenport mentioned that no special triple clock latching is necessary with your register  settings).  
@JamesCWhite little kicks for now ... but it looks promising :)

**Comment by bassetware** [comment]
Ich had issues with blank or non working panles. I added some Level Stifters and Afterwards the Panel worked with an esp32. Esp8266 (d1 Mini) does Not Work though. 

Support from AliExpress says that 5v Logiclevel ist needed.

**Comment by DaveDavenport** [comment]
@bassetware  Do they use the the other chip? (see earlier in issue) I think the raspberry pi I used, also is only 3.3V. (and that worked).

**Comment by DaveDavenport** [comment]
Some observations with the above code.

First block of leds do not light correctly (also in above video).

Panel uses an absurd amount of current compared to rbpi version.. 2 lines white is around 4.4Amp. Panel is getting hot.

Played a bit with the extra cycles during the latch, makes no difference.

**Comment by DaveDavenport** [comment]
I have it working.  But it ain't great.
I had to ditch the hardware SPI, because you need to raise Latch 3 clock cycles for the last  pixel.

**Comment by 2dom** [comment]
Wanna share something so we can have a look?

**Comment by DaveDavenport** [comment]
yeah, it was 3 at night. will post later :stuck_out_tongue_closed_eyes:

**Comment by 2dom** [comment]
:) We might be able to use spi for all but the last byte which is not ideal but fair enough

**Comment by DaveDavenport** [comment]
I am having problem that loop is called only once, and somehow it does not get past delay() . I am taking to much time in interrupt.

```diff
diff --git a/PxMatrix.h b/PxMatrix.h
index 9f7723a..c6d2600 100644
--- a/PxMatrix.h
+++ b/PxMatrix.h
@@ -51,10 +51,8 @@ BSD license, check license.txt for more information
 #endif
 
 // HW SPI PINS
-#define SPI_BUS_CLK 14
-#define SPI_BUS_MOSI 13
-#define SPI_BUS_MISO 12
-#define SPI_BUS_SS 4
+#define SPI_BUS_CLK  D5
+#define SPI_BUS_MOSI D7
 
 // Either the panel handles the multiplexing and we feed BINARY to A-E pins
 // or we handle the multiplexing and activate one of A-D pins (STRAIGHT)
@@ -313,9 +311,9 @@ inline void PxMATRIX::writeRegister(uint16_t reg_value, uint8_t reg_position)
     // b13  - 9   =1 screen on
     // b13  - 6   =1 screen off
 
-    digitalWrite(SPI_BUS_CLK,HIGH); // CCK LOW
-    digitalWrite(_OE_PIN,LOW);
-    digitalWrite(_LATCH_PIN,HIGH);
+    digitalWrite(SPI_BUS_CLK,LOW); // CCK LOW
+    digitalWrite(_OE_PIN,HIGH);
+    digitalWrite(_LATCH_PIN,LOW);
     digitalWrite(_A_PIN,HIGH);
     digitalWrite(_B_PIN,LOW);
     digitalWrite(_C_PIN,LOW);
@@ -337,10 +335,10 @@ inline void PxMATRIX::writeRegister(uint16_t reg_value, uint8_t reg_position)
       delay(1);
       if ((bit_counter ==  (_send_buffer_size*8 - reg_position-1)))
       {
-        digitalWrite(_LATCH_PIN,LOW);
+        digitalWrite(_LATCH_PIN,HIGH);
       }
     }
-    digitalWrite(_LATCH_PIN,HIGH);
+    digitalWrite(_LATCH_PIN,LOW);
   }
   digitalWrite(_OE_PIN,HIGH);
 
@@ -647,17 +645,8 @@ void PxMATRIX::begin(uint8_t row_pattern) {
   _row_sets_per_buffer = _rows_per_buffer/_row_pattern;
   _send_buffer_size=_pattern_color_bytes*3;
 
-#ifdef ESP8266
-  SPI.begin();
-#endif
-#ifdef ESP32
-  SPI.begin(SPI_BUS_CLK, SPI_BUS_MISO, SPI_BUS_MOSI, SPI_BUS_SS);
-#endif
-
-  SPI.setDataMode(SPI_MODE0);
-  SPI.setBitOrder(MSBFIRST);
-  SPI.setFrequency(20000000);
-
+  pinMode ( SPI_BUS_CLK, OUTPUT);
+  pinMode ( SPI_BUS_MOSI, OUTPUT);
   pinMode(_OE_PIN, OUTPUT);
   pinMode(_LATCH_PIN, OUTPUT);
   pinMode(_A_PIN, OUTPUT);
@@ -676,7 +665,7 @@ void PxMATRIX::begin(uint8_t row_pattern) {
     pinMode(_D_PIN, OUTPUT);
     digitalWrite(_D_PIN, LOW);
   }
-  if (_row_pattern >=32)
+  if (_row_pattern > 32)
   {
     pinMode(_E_PIN, OUTPUT);
     digitalWrite(_E_PIN, LOW);
@@ -800,54 +789,39 @@ void PxMATRIX::display(uint16_t show_time) {
 #endif
   for (uint8_t i=0;i<_row_pattern;i++)
   {
-    if (_fast_update && (_brightness==255)){
-
-      // This will clock data into the display while the outputs are still
-      // latched (LEDs on). We therefore utilize SPI transfer latency as LED
-      // ON time and can reduce the waiting time (show_time). This is rather
-      // timing sensitive and may lead to flicker however promises reduced
-      // update times and increased brightness
-
-      set_mux((i+_row_pattern-1)%_row_pattern);
-      digitalWrite(_LATCH_PIN,HIGH);
-      digitalWrite(_OE_PIN,0);
-      start_time = micros();
+      set_mux(i);
+      for ( int x = 0; x < _send_buffer_size-1; x++ ) {
+          shiftOut(SPI_BUS_MOSI, SPI_BUS_CLK, MSBFIRST, PxMATRIX_buffer[_display_color][i*_send_buffer_size+x]);
+          //SPI.writeBytes(&PxMATRIX_buffer[_display_color][i*_send_buffer_size],_send_buffer_size);
+      }
+      uint8_t v =  PxMATRIX_buffer[_display_color][i*_send_buffer_size+_send_buffer_size-1];
+      for ( int b =0; b < 8; b++ ){
+          digitalWrite(SPI_BUS_MOSI, ((v>>b)&1) == 1);
+          digitalWrite(SPI_BUS_CLK,HIGH);
+          delayMicroseconds(1);
+          digitalWrite(SPI_BUS_CLK,LOW);
+          delayMicroseconds(1);
+
+          if (b == (8-1-3))
+              digitalWrite(_LATCH_PIN,HIGH);
+      }
       digitalWrite(_LATCH_PIN,LOW);
-      delayMicroseconds(1);
-
-#ifdef double_buffer
-      SPI.writeBytes(&PxMATRIX_buffer[_display_color][buffer_size*_active_buffer+i*_send_buffer_size],_send_buffer_size);
-#else
-      SPI.writeBytes(&PxMATRIX_buffer[_display_color][i*_send_buffer_size],_send_buffer_size);
-#endif
-      while ((micros()-start_time)<show_time)
-        delayMicroseconds(1);
+      digitalWrite(_OE_PIN,0); //<<<< insert this
+      delayMicroseconds(show_time);
       digitalWrite(_OE_PIN,1);
-    }
-    else
-    {
-      set_mux(i);
-#ifdef double_buffer
-      SPI.writeBytes(&PxMATRIX_buffer[_display_color][buffer_size*_active_buffer+i*_send_buffer_size],_send_buffer_size);
-#else
-      SPI.writeBytes(&PxMATRIX_buffer[_display_color][i*_send_buffer_size],_send_buffer_size);
-#endif
-      latch(show_time*(uint16_t)_brightness/255);
-    }
+      //latch(show_time*(uint16_t)_brightness/255);
   }
   _display_color++;
   if (_display_color>=PxMATRIX_COLOR_DEPTH)
   {
     _display_color=0;
-#ifdef double_buffer
-    _active_buffer=_selected_buffer;
-#endif
   }
 }
 
 void PxMATRIX::flushDisplay(void) {
-  for (int ii=0;ii<_send_buffer_size;ii++)
-    SPI.write(0x00);
+  for (int ii=0;ii<_send_buffer_size;ii++) {
+      shiftOut(SPI_BUS_MOSI, SPI_BUS_CLK, MSBFIRST, 0x00);
+  }
 }
 
 void PxMATRIX::displayTestPattern(uint16_t show_time) {
```

This for me makes the display function work (not test patterns! need to change that code still).

Initialization is not perfect, but works most of the time.

**Comment by DaveDavenport** [comment]
With this I got a (flickery) MorphingClock working.

**Comment by DaveDavenport** [comment]
vid: https://filebin.pw/axW/

**Comment by DaveDavenport** [comment]
```diff
diff --git a/PxMatrix.h b/PxMatrix.h
index 9f7723a..02450ac 100644
--- a/PxMatrix.h
+++ b/PxMatrix.h
@@ -51,10 +51,8 @@ BSD license, check license.txt for more information
 #endif
 
 // HW SPI PINS
-#define SPI_BUS_CLK 14
-#define SPI_BUS_MOSI 13
-#define SPI_BUS_MISO 12
-#define SPI_BUS_SS 4
+#define SPI_BUS_CLK  D5
+#define SPI_BUS_MOSI D7
 
 // Either the panel handles the multiplexing and we feed BINARY to A-E pins
 // or we handle the multiplexing and activate one of A-D pins (STRAIGHT)
@@ -313,9 +311,9 @@ inline void PxMATRIX::writeRegister(uint16_t reg_value, uint8_t reg_position)
     // b13  - 9   =1 screen on
     // b13  - 6   =1 screen off
 
-    digitalWrite(SPI_BUS_CLK,HIGH); // CCK LOW
-    digitalWrite(_OE_PIN,LOW);
-    digitalWrite(_LATCH_PIN,HIGH);
+    digitalWrite(SPI_BUS_CLK,LOW); // CCK LOW
+    digitalWrite(_OE_PIN,HIGH);
+    digitalWrite(_LATCH_PIN,LOW);
     digitalWrite(_A_PIN,HIGH);
     digitalWrite(_B_PIN,LOW);
     digitalWrite(_C_PIN,LOW);
@@ -337,10 +335,10 @@ inline void PxMATRIX::writeRegister(uint16_t reg_value, uint8_t reg_position)
       delay(1);
       if ((bit_counter ==  (_send_buffer_size*8 - reg_position-1)))
       {
-        digitalWrite(_LATCH_PIN,LOW);
+        digitalWrite(_LATCH_PIN,HIGH);
       }
     }
-    digitalWrite(_LATCH_PIN,HIGH);
+    digitalWrite(_LATCH_PIN,LOW);
   }
   digitalWrite(_OE_PIN,HIGH);
 
@@ -647,17 +645,8 @@ void PxMATRIX::begin(uint8_t row_pattern) {
   _row_sets_per_buffer = _rows_per_buffer/_row_pattern;
   _send_buffer_size=_pattern_color_bytes*3;
 
-#ifdef ESP8266
-  SPI.begin();
-#endif
-#ifdef ESP32
-  SPI.begin(SPI_BUS_CLK, SPI_BUS_MISO, SPI_BUS_MOSI, SPI_BUS_SS);
-#endif
-
-  SPI.setDataMode(SPI_MODE0);
-  SPI.setBitOrder(MSBFIRST);
-  SPI.setFrequency(20000000);
-
+  pinMode ( SPI_BUS_CLK, OUTPUT);
+  pinMode ( SPI_BUS_MOSI, OUTPUT);
   pinMode(_OE_PIN, OUTPUT);
   pinMode(_LATCH_PIN, OUTPUT);
   pinMode(_A_PIN, OUTPUT);
@@ -676,7 +665,7 @@ void PxMATRIX::begin(uint8_t row_pattern) {
     pinMode(_D_PIN, OUTPUT);
     digitalWrite(_D_PIN, LOW);
   }
-  if (_row_pattern >=32)
+  if (_row_pattern > 32)
   {
     pinMode(_E_PIN, OUTPUT);
     digitalWrite(_E_PIN, LOW);
@@ -800,54 +789,47 @@ void PxMATRIX::display(uint16_t show_time) {
 #endif
   for (uint8_t i=0;i<_row_pattern;i++)
   {
-    if (_fast_update && (_brightness==255)){
-
-      // This will clock data into the display while the outputs are still
-      // latched (LEDs on). We therefore utilize SPI transfer latency as LED
-      // ON time and can reduce the waiting time (show_time). This is rather
-      // timing sensitive and may lead to flicker however promises reduced
-      // update times and increased brightness
-
-      set_mux((i+_row_pattern-1)%_row_pattern);
-      digitalWrite(_LATCH_PIN,HIGH);
-      digitalWrite(_OE_PIN,0);
-      start_time = micros();
+      set_mux(i);
+      for ( int x = 0; x < _send_buffer_size-1; x++ ) {
+          uint8_t v =  PxMATRIX_buffer[_display_color][i*_send_buffer_size+x];
+          for ( int b =0; b < 8; b++ ){
+              if ( ((v>>(7-b))&1) == 1)
+                  GPIO_REG_WRITE(GPIO_OUT_W1TS_ADDRESS, 1<<SPI_BUS_MOSI);
+              else 
+                  GPIO_REG_WRITE(GPIO_OUT_W1TC_ADDRESS, 1<<SPI_BUS_MOSI);
+              GPIO_REG_WRITE(GPIO_OUT_W1TS_ADDRESS, 1<<SPI_BUS_CLK);
+              GPIO_REG_WRITE(GPIO_OUT_W1TC_ADDRESS, 1<<SPI_BUS_CLK);
+          }
+      }
+      uint8_t v =  PxMATRIX_buffer[_display_color][i*_send_buffer_size+_send_buffer_size-1];
+      for ( int b =0; b < 8; b++ ){
+          if ( ((v>>(7-b))&1) == 1)
+              GPIO_REG_WRITE(GPIO_OUT_W1TS_ADDRESS, 1<<SPI_BUS_MOSI);
+          else 
+              GPIO_REG_WRITE(GPIO_OUT_W1TC_ADDRESS, 1<<SPI_BUS_MOSI);
+          GPIO_REG_WRITE(GPIO_OUT_W1TS_ADDRESS, 1<<SPI_BUS_CLK);
+          GPIO_REG_WRITE(GPIO_OUT_W1TC_ADDRESS, 1<<SPI_BUS_CLK);
+
+          if (b == (8-1-3))
+              digitalWrite(_LATCH_PIN,HIGH);
+      }
       digitalWrite(_LATCH_PIN,LOW);
-      delayMicroseconds(1);
-
-#ifdef double_buffer
-      SPI.writeBytes(&PxMATRIX_buffer[_display_color][buffer_size*_active_buffer+i*_send_buffer_size],_send_buffer_size);
-#else
-      SPI.writeBytes(&PxMATRIX_buffer[_display_color][i*_send_buffer_size],_send_buffer_size);
-#endif
-      while ((micros()-start_time)<show_time)
-        delayMicroseconds(1);
+      digitalWrite(_OE_PIN,0); //<<<< insert this
+      delayMicroseconds(show_time);
       digitalWrite(_OE_PIN,1);
-    }
-    else
-    {
-      set_mux(i);
-#ifdef double_buffer
-      SPI.writeBytes(&PxMATRIX_buffer[_display_color][buffer_size*_active_buffer+i*_send_buffer_size],_send_buffer_size);
-#else
-      SPI.writeBytes(&PxMATRIX_buffer[_display_color][i*_send_buffer_size],_send_buffer_size);
-#endif
-      latch(show_time*(uint16_t)_brightness/255);
-    }
+      //latch(show_time*(uint16_t)_brightness/255);
   }
   _display_color++;
   if (_display_color>=PxMATRIX_COLOR_DEPTH)
   {
     _display_color=0;
-#ifdef double_buffer
-    _active_buffer=_selected_buffer;
-#endif
   }
 }
 
 void PxMATRIX::flushDisplay(void) {
-  for (int ii=0;ii<_send_buffer_size;ii++)
-    SPI.write(0x00);
+  for (int ii=0;ii<_send_buffer_size;ii++) {
+      shiftOut(SPI_BUS_MOSI, SPI_BUS_CLK, MSBFIRST, 0x00);
+  }
 }
 
 void PxMATRIX::displayTestPattern(uint16_t show_time) {
```

This one can run around 10 times faster then the above version! less of a headache to use :stuck_out_tongue:

**Comment by DaveDavenport** [comment]
Ill try to post a cleaner patch tomorrow. Made some more tiny improvements.

**Comment by DaveDavenport** [comment]
https://filebin.pw/E6bN/

**Comment by 2dom** [comment]
Great...could you just create a pull request?

**Comment by DaveDavenport** [comment]
I will see. Not sure I have the time.

**Comment by JamesCWhite** [comment]
@DaveDavenport Did you utilize the _E_ input on your panel? My panel pinout is unmarked and I currently have _E_ tied to gnd. I tried my hand at making your edits but upon running the pattern test I only get these 8 white dots. I'm thinking either I borked up your edits or maybe I need the _E_ input after all.
![img_20190119_110806312](https://user-images.githubusercontent.com/6243304/51429283-28ee3b80-1bdb-11e9-9ba4-3045103ba178.jpg)

**Comment by Bastelschlumpf** [comment]
I connected all the pins also the E pin and set the driver chip to the FM6124 in my pull request version.


#ifdef ESP8266

#include <Ticker.h>
Ticker display_ticker;
#define P_LAT 16
#define P_A 5
#define P_B 4
#define P_C 15
#define P_D 12
#define P_E 0
#define P_OE 2

#endif

// Pins for LED MATRIX
PxMATRIX display(64, 32, P_LAT, P_OE, P_A, P_B, P_C, P_D, P_E);

void setup() {
  Serial.begin(9600);
  display.setDriverChip(FM6124);
  display.begin(16);

**Comment by JamesCWhite** [comment]
> I connected all the pins also the E pin and set the driver chip to the FM6124 in my pull request version.

I am using the 6126A and I had no real luck so far with your pull request or the update by DaveDavenport. Probably means I'm doing it wrong lol

**Comment by 2dom** [comment]
Hi guys ... good work! I merged the patch to master (still want to get this working with hardware SPI but ok for now). 

I had a minute to try it on an esp8266 based nodeMCU and I noticed that it locks the MCU (probably due to too much time spend in the update call). Are you guys on Esp32 or esp8266?

**Comment by DaveDavenport** [comment]
Yes it works for me with esp8266.

I tried enabling SPI again (not even using it), but that somehow screws communication up.  Need to attach scope/logic analyzer to see what happens.

**Comment by DaveDavenport** [comment]
Scope indicates that SCK is around 3.3 MHz.

**Comment by 2dom** [comment]
OK ... I spend some time logic analyzing....problem was that all SPI related ports are set to pinMode special in SPI.begin. Setting them to pinMode output before the bitbanging part does the trick. Now hardware SPI works and we only use bitbanging for the last byte. (check latest master)

Glad that we got it working now but these chips are terrible - I would not advise to buy panels that use them. FastUpdate will not work - hence you will probably run into flicker problems with chaining.

**Comment by 2dom** [comment]
Special thanks to  @shades66 @DaveDavenport @Bastelschlumpf !

**Comment by DaveDavenport** [comment]
@2dom  great!, will test it.

**Comment by DaveDavenport** [comment]
@2dom should https://github.com/2dom/PxMatrix/blob/master/PxMatrix.h#L877 not be on line 852. (before the spi writing)

and should probably remove the delay(s) from writeRegister. It slows startup down a lot.

**Comment by DaveDavenport** [comment]
> Glad that we got it working now but these chips are terrible - I would not advise to buy panels that use them. FastUpdate will not work - hence you will probably run into flicker problems with chaining.

I am unsure why they made it like this, and not maintain backwards compatibility. Should have been easy enough. (and still have config registers)

**Comment by 2dom** [comment]
@DaveDavenport why do you want to move 
`uint8_t v = PxMATRIX_buffer[_display_color][i*_send_buffer_size + _send_buffer_size - 1];`
 up? (Note ... I just pushed some changes so the line numbers don't line up anymore).

**Comment by 2dom** [comment]
> > Glad that we got it working now but these chips are terrible - I would not advise to buy panels that use them. FastUpdate will not work - hence you will probably run into flicker problems with chaining.
> 
> I am unsure why they made it like this, and not maintain backwards compatibility. Should have been easy enough. (and still have config registers)

Yep ... and this counting the data clock edges for different commands (especially latch) is just the most stupid idea ever.

**Comment by DaveDavenport** [comment]
aah , github does not fix that :smile:  I was pointing to the set_mux(i) line.
To the beginning of the  block, (before spi transfer/pimode) so it is consistent to the other  block.

**Comment by DaveDavenport** [comment]
> > > Glad that we got it working now but these chips are terrible - I would not advise to buy panels that use them. FastUpdate will not work - hence you will probably run into flicker problems with chaining.
> > 
> > 
> > I am unsure why they made it like this, and not maintain backwards compatibility. Should have been easy enough. (and still have config registers)
> 
> Yep ... and this counting the data clock edges for different commands (especially latch) is just the most stupid idea ever.

Still need to lookup how the raspberry pi version does it. (as it only needed the config register update).

**Comment by houk001** [comment]
Its working !!!! Some issues still in 80 MHz but on 160MHz MorphingClock running nice.

**Comment by PeterTeunissen** [comment]
Same for me! Wemos d1 mini pro on 160MHz works with these latest changes!!

Thanks sooooo much to the guys that actually understand these boards and use their own time to provide this library!

**Comment by iSmile67** [comment]
Thanks for the great work you have done. Finally it worked. Does everyone have a flickering on 6124? Or should I do everything not on the wires but on the pcb?

**Comment by 2dom** [comment]
Did you use a refresh/update cycle time of 2ms and a show time of 30?On 21 Jan 2019 07:35, iSmile67 <notifications@github.com> wrote:Thanks for the great work you have done. Finally it worked. Does everyone have a flickering on 6124? Or should I do everything not on the wires but on the pcb?

**Comment by iSmile67** [comment]
> Did you use a refresh/update cycle time of 2ms and a show time of 30?

Not. Just used your library and added lines to the MorphingClock sketch
```
 display.setDriverChip(FM6124);
  display.begin(16);
  display.setFastUpdate(true);
```
Slight flicker on video [video](https://vk.com/lisovskiyan?z=video7685647_456239047%2F8a992be2dbb5dc53f3%2Fpl_wall_7685647 - )

**Comment by DaveDavenport** [comment]
After some testing of the code I noticed that one blue led of one pixel is broken on my panel, so I contacted the seller.   I asked him if he still ships panels with the old driver, will keep you guys informed on the answer..

**Comment by carlhako** [comment]
Could be my panel but I dont get any green using the following code. I get red, blue then nothing.

`display.fillRect(0,0,64,32,display.color565(255, 0, 0));
  delay(1000);
  display.fillRect(0,0,64,32,display.color565(0, 255, 0));
  delay(1000);
  display.fillRect(0,0,64,32,display.color565(0, 0, 255));
  delay(1000);`

**Comment by mshill** [comment]
I have one issue when trying to compile for ESP32 and using the FM62126A chips:

`PxMatrix.h:882:24: error: 'GPIO_OUT_W1TC_ADDRESS' was not declared in this scope

         GPIO_REG_WRITE(GPIO_OUT_W1TC_ADDRESS, 1 << SPI_BUS_CLK);
`
Any idea what I'm doing wrong? I searched this one exhaustively and seems GPIO_OUT_W1TC_ADDRESS and GPIO_OUT_W1TS_ADDRESS are not defined for ESP32...

I was excited when I ran across this thread, I replaced two of these things with the same results... :|
Thanks for EVERYONE's work.

**Comment by houk001** [comment]
I tried also MorphingClockRemix and second MorphingClockRemixRemix. Both have high problem to work. Freezing, esp8266 hang, mixed colors, pixel lighting up that should be off. I think that there is still efficiency problem with this new chips. Maybe there is something to speedup a little.

**Comment by 2dom** [comment]
@mshill Try again with the latest master?

**Comment by mshill** [comment]
Well, at least it compiled! Still didn't work with the P3 FM6126A units, but I have an ESP32 running the pixeltime sketch on an older P4. I think it is time to pack these P3 units up and forget them...

Thanks for the help, This is a really helpful group of folks!

Now to work through the other ESP32 compile errors on MorphingClock... :(

**Comment by fotosettore** [comment]
> I think it is time to pack these P3 units up and forget them...
there are dozens of users that bought p3 with fm6126a on board, me too .
I opened this talk to look for a solution, not to pack the panels.
These will be the only units that surely will be sold in the future.
"back to the past" in not a solution ...

**Comment by DaveDavenport** [comment]
Got a reply from the alie seller:
```
now we don't sell FM6124 and FM6126A chip led module, because some customers said it is difficult to use. Now our chip is ICN2037, it is easier to use. 
```

**Comment by DaveDavenport** [comment]
Got a 2nd message that all order after 10th of January should have the 'simple' chip again.
(SRYLED Display Store)

**Comment by fotosettore** [comment]
sounds great !!!

but ... for all panels bought before ?
and panels bought from another seller ?
and again : for the future ?

what are we doing ? do you throw all the work done so far?
I'm not afraid of losing $ 200 of 10 unusable panels.
I'm afraid to buy others with the same problems, because I'm sure that other makers will sooner or later have the problem of the FM6124 and FM6126A chip led module.
if they have put on sale FM6124 panels means that sooner or later we will find them in the middle of the feet :-)))
do not you think it's possible ?

**Comment by DaveDavenport** [comment]
To be honest I would like to get proper documentation for the chip. (as I said, the rbpi version worked with setting the registers and nothing more).

Maybe there is a way to get more decent behavior.

**Comment by fotosettore** [comment]
thanks to all those who will find a solution, which in any case (unfortunately) is not just about me.

**Comment by DaveDavenport** [comment]
I checked (quickly) the code of the rbpi one, there they strobe after loading the data. Curious why it works for them (tm).

**Comment by DaveDavenport** [comment]
(sorry for the ramblings).
tried to reproduce the 'switching' pattern from https://github.com/hzeller/rpi-rgb-led-matrix.
Did not manage to get it to work (I get nice white vertical stripes).

I wonder why it works on pi without the 3 cycle latch, but not on the esp.  Regrettable I don't have a fast enough logic analyzer to do a good comparison.

**Comment by 2dom** [comment]
Thanks DaveDavenport ... could you open a new issue with something like "FM6124 performance optimizations" and a pull request ? Thanks!

Closing this issue now since basic support is done and this is discussion is getting too long

**Comment by 2dom** [comment]
@DaveDavenport any news regarding the RPI code ?

**Comment by DaveDavenport** [comment]
@dom I am very busy atm, I hope to have some time in 2 weeks ago. I have several cleanup patches (speed up the writeRegister and fix  some other stuff).

**Comment by DaveDavenport** [comment]
Got new panel in today, it works correctly with the old code :smile:

**Comment by abricier** [comment]
Hi,
I bought a panel after/during the "FM62126A Gate" and got a model using the 'chipon' chip instead of the FM62126 or FM62126A and after the 10th of january. So I was wondering, @DaveDavenport, if your last test with your newest panel was on a FM62126 or on a 'chipon' version.

For future reference, here is the message for the seller (2019/01/14):
"Hello friend, we don't sell FM chip now, we use chipone 2037chip, it can work with asynchronous card."

**Comment by DaveDavenport** [comment]
I think it is a chipone chip. Seems to work ok.

**Comment by DaveDavenport** [comment]
Ugh this is frustrating.. 2 out of 3 panels arrived damaged. 
This is the last shipment. :(  sometimes I hate the post office.
![pr3-4](https://user-images.githubusercontent.com/406245/52296646-321f2c80-297f-11e9-9191-529f43e9a076.jpg)

**Comment by 2dom** [comment]
That sucks ... how do they manage to do that? Mine usually come in 5 tonnes of bubble wrap.
Did you have time to look into the code again?

**Comment by DaveDavenport** [comment]
throwing around in post I guess. I managed to fix it for now.

I am waiting for 2 china packages, then I will continue again.

**Comment by bobdavis321** [comment]
Here is the code to fix the newer panels with an Arduino: http://bobdavis321.blogspot.com/2019/02/p3-64x32-hub75e-led-matrix-panels-with.html

**Comment by IAmOrion** [comment]
Hi There,

I purchased a 64x32 panel off Amazon.co.uk the other day - and after hours of research I came across this post, and mine is indeed the FM6126A variety! 

I finally managed to get it to work... (I was originally going to post saying it's still not working me, but my 10A power supply was actually outputting 5.20 volts from when I had previously used it on my Pi (to get rid of the lightning bolt icon that appears)) and apparently, the extra voltage screws up the clock or something.  Only when I dialled it in to 5v exactly did it finally work! 

That being said, I still have an issue, and wanted to ask if anyone knew if this was a fault with the display or if I need an additional step in my code.  Here is my results of displaying a picture (It's easier to see the problem this way):

![IMG_0154](https://user-images.githubusercontent.com/3501380/56014210-30117b00-5cec-11e9-8bda-efe6c41b9930.jpeg)

The top left quarter is not as bright as the rest of the display :( 

Have I missed a step out code wise, of is this potentially a defective matrix?

Kind regards

**Comment by bobdavis321** [comment]
That one panel is not properly reset.

**Comment by IAmOrion** [comment]
> That one panel is not properly reset.

How do I ensure it resets properly.  When I looked at this again this morning, when I first powered it up, the pixel area had "Changed location"..... in other words, the dim area seen in the pic were now bright and correct, and instead a different area was dim.  Nothing changed, I didn't change or reupload code, I simply plugged it back in to look again and noticed different parts weren't working this time.

these new FM6126A based ones seem to be a real PITA :(

**Comment by bobdavis321** [comment]
Check your grounds, you might have too much current resulting is a voltage loss and data not being sent properly.  Also I hope you have at least two power supplies capable of 10 amps to run that many panels.

**Comment by IAmOrion** [comment]
> Check your grounds, you might have too much current resulting is a voltage loss and data not being sent properly. Also I hope you have at least two power supplies capable of 10 amps to run that many panels.

I've already checked my grounds, they're fine (so far as I can tell).  Voltage is still 5.0v even when all powered up.  Power Supply is a 5v 10a supply.

Also, it's just one panel - it's a single "64x32" panel.

**Comment by Robiv8** [comment]
@IAmOrion 
Can you please send me the Sketch what you used.
Thanks

**Comment by bobdavis321** [comment]
Are you using an Arduino or a Raspberry Pi?  The Arduino code is on my blog
at bobdavis321.blogspot.com

**Comment by IAmOrion** [comment]
> @IAmOrion
> Can you please send me the Sketch what you used.
> Thanks

@Robiv8 Attached

[32x64_LED_Matrix_Demo.zip](https://github.com/2dom/PxMatrix/files/3091479/32x64_LED_Matrix_Demo.zip)

**Comment by Robiv8** [comment]
Thank you bobdavis and also IAmOrion

**Comment by Robiv8** [comment]
i have this error wenn compile
Whats wrong?

> 32x64_LED_Matrix_Demo:189:25: error: 'FM6126A' was not declared in this scope
>    display.setDriverChip(FM6126A); // THIS IS THE IMPORTANT LINE FOR GETTING MATRIXES WITH THE FM6126A CHIP WORKING!!!
>                          ^
> exit status 1
> 'class PxMATRIX' has no member named 'setDriverChip'

**Comment by fotosettore** [comment]
the command **display.setBrightness** seems not working in the example "32x64_LED_Matrix_Demo.zip"

**Comment by IAmOrion** [comment]
> the command **display.setBrightness** seems not working in the example "32x64_LED_Matrix_Demo.zip"

My sketch doesn't even have **display.setBrightness** in it!!  Can you screenshot your error please?

Also, are you using the PxMatrix library downloaded here (via Github)? The one in the IDE is outdated

**Comment by fotosettore** [comment]
no error ... simply it does not work !
display.setBrightness(10) or display.setBrightness(255) has no effect.
in the "normal" panels (not FM6126A) the command works fine.
The library was download here

**Comment by IAmOrion** [comment]
> no error ... simply it does not work !
> display.setBrightness(10) or display.setBrightness(255) has no effect.
> in the "normal" panels (not FM6126A) the command works fine.
> The library was download here

I'm still confused - You're talking about the zip file I uploaded in this post (https://github.com/2dom/PxMatrix/issues/53#issuecomment-484253208) right?

That's specifically for the FM6126A - although to be honest, you only really need to comment **out** this line: display.setDriverChip(FM6126A); and it should work for any.

I have to repeat though - no where in my Sketch do I use "display.setBrightness" - but it is apart of the PxMatrix.h file in the library:

  // Set the brightness of the panels
  inline void setBrightness(uint8_t brightness);

Since my demo sketch does **NOT** use display.setBrightness anywhere - I have to ask again, where you are seeing this? - Or do you mean, you have **ADDED** some code to my demo/example, and **YOU** added the setBrightness, but it's not having any effect?? If that's the case, you should perhaps open a new Issue with the PxMatrix library itself as it's nothing to do with my code

**Comment by fotosettore** [comment]
i downloaded your example of #53 and it works fine togheter FM6126A and ICN2037
in ICN, if i put command display.setBrightness it works fine
in FM does not work
the library is the same. i changed only panel
display.setBrightness is inside pxmatrix of domenic

**Comment by IAmOrion** [comment]
> i downloaded your example of #53 and it works fine togheter FM6126A and ICN2037
> in ICN, if i put command display.setBrightness it works fine
> in FM does not work
> the library is the same. i changed only panel
> display.setBrightness is inside pxmatrix of domenic
> asap i will post a example video

Right, Now I understand - you'll need to create a new issue and query the creator of PxMatrix - I assume it's something to do with the FM6126A chips that perhaps doesn't use the setBrightness function.

**Comment by 2dom** [comment]
Use ESP32 for best results with these panels and set `display.setDriverChip(FM6124);`
